### PR TITLE
ModularTokenizer save path

### PR DESCRIPTION
**Pull request by floccinauc**
A ModularTokenizer now has load() and save() functions that only receive a single load/save path where all modular sub-tokenizer jsons are stored alongside a config.yaml that specified their relations and parameters. The config.yaml is automatically generated by save(). The save files can be moved to a different directory, and load() will still function, as long as the new path is passed to it.
This is opposed to the previous functionality of load_from_jsons and save_jsons that used a config file where relevant sub-tokenizer paths, parameters and relations had to be specified.

### CODE REVIEW

File: fusedrug/data/tokenizer/modulartokenizer/README.md [hunk 1]
```diff
@@ -10,8 +10,12 @@ Creates a modular tokenizer that combines multiple initial tokenizers, adjusting
 * They all share the same special tokens (combined special tokens from all the source tokenizers),
 * Each tokenizer retains its regular tokens, however their IDs are remapped to a single space, with no overlaps.
 Note: If a token has the same meaning across all input types (e.g. special tokens, like SEP, EOS, sentinel tokens), it should be defined as special token in at least one of the initial (input) tokenizers.
+### save():
+Saves all mudular tokenizer information to a given path.
 ### save_jsons():
```
Comment (mosheraboh) [inline fusedrug/data/tokenizer/modulartokenizer/README.md:15]: Do we want to keep both save and save_jsons?
If you only recommend to use save() we can remove save_jsons from README.

File: fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py [hunk 2]
```diff
@@ -476,6 +528,75 @@ def save_jsons(self, tokenizers_info: Optional[List] = None) -> None:
                     os.makedirs(os.path.dirname(out_path))
                 tokenizer_inst.save(out_path)
 
+    def save(self, path: str) -> None:
+        """Saves all information needed to reconstruct the modular tokenizer to path.
+        After saving, path will contain the following:
+        - json files: modular json files (i.e. that have common special tokens, and that all map to consistent ID space)
+        - config.yaml: a config file that contains the tokenizer_info list with information from self.tokenizer_info, defining the relations between the different tokenizers.
+        ** Since all modular tokenizers are found in the same path, and the path may change when it's passed between users, the path is not included in the config
```
Comment (mosheraboh) [inline fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py:536]: Great

File: fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py [hunk 3]
```diff
@@ -476,6 +528,75 @@ def save_jsons(self, tokenizers_info: Optional[List] = None) -> None:
...
+                input_json_path=self.tokenizers_info[t_type]["json_path"],
+                base_path=None,
+            )
+            tokenizers_info_cfg = set_field(
+                tokenizers_info_cfg=tokenizers_info_cfg,
+                name=t_type,
+                key="modular_json_path",
+                val=config_out_path,
+            )
+            # Original json path (for the json of the tokenizer used to create the original instance of this ModularTokenizer) is no longer relevant,
```
Comment (mosheraboh) [inline fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py:587]: Maybe set to None or "N/A" to better reflect it?
Another option is to copy the original one as well.

File: fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py [hunk 4]
```diff
@@ -284,6 +295,47 @@ def load_from_jsons(tokenizers_info: List) -> Any:
         """
         return ModularTokenizer(tokenizers_info=tokenizers_info, load_adjusted_jsons=True)
 
+    @staticmethod
+    def load(path: str) -> Any:
```
Comment (mosheraboh) [inline fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py:299]: is it the method name in huggingface? is it equivalent to from_file?

File: fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py [hunk 5]
```diff
@@ -284,6 +295,47 @@ def load_from_jsons(tokenizers_info: List) -> Any:
...
+        """
+
+        def fix_json_paths(
+            loaded_conf: omegaconf.listconfig.ListConfig,
+            path: str,
+        ) -> omegaconf.listconfig.ListConfig:
+            """Since the path passed to ModularTokenizer.load() must contain all of the modular jsons and the config that defines their relations, all json
+            paths in the config must point to path. This function replaces any dirname of any json in the loaded config (found in path) with path.
+            Args:
+                loaded_conf (Union[List, omegaconf.listconfig.ListConfig]): _description_
```
Comment (mosheraboh) [inline fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py:322]: I guess that _description_ is auto-generated for you?

File: fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py [hunk 6]
```diff
@@ -284,6 +295,47 @@ def load_from_jsons(tokenizers_info: List) -> Any:
...
+                path (str): _description_
```
Comment (mosheraboh) [inline fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py:323]: path to a dir that contains config.yaml, right?

**Review by mosheraboh — APPROVED** [review]
Looks great!
Some minor thoughts/comments inline

File: fusedrug/data/tokenizer/modulartokenizer/README.md (hunk 1 above)
Comment (floccinauc) [reply fusedrug/data/tokenizer/modulartokenizer/README.md:15]: Removed from readme and commented in the examples

File: fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py (hunk 3 above)
Comment (floccinauc) [reply fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py:587]: Done and tested that it works

File: fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py (hunk 6 above)
Comment (floccinauc) [reply fusedrug/data/tokenizer/modulartokenizer/multi_tokenizer.py:323]: yes

**Pull request merged**